### PR TITLE
Improve formatting for React Hooks call

### LINKED ISSUE #7671: Ugly React Hooks
[Playground Link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAgCU4BDMGAGn2HwFcBnOAZRhPiobgFk50J8AvvgBmOCOnwAdEDlLlpAbklRlkKPQKYSORgDFaAGwMA5Eujj4AvCMMmzFywD5qy-PjUb8Ad1wATelY2RqbmAHQwOBgAFACUofSYBqgwUQD0kpL0ANSpccKoBvA4UQBCEBAGpFAxSlBuHgT5OjAhDt5+9PEAFqjCKTWu7tCeBiQarYE+OP6hAFYQqFBR0lIgA3X4sjC0OHU0TeP2VKOH5oK1AsrKwrRQ5KjQ+ACCmJixLhsN+ADaN8FH+EYMH0-3MAF1ApwWGw4MsQNJ1vVhgR9qhmq1jmMWvZBJDGDw+MsNm53k58FodHAQXZzFE-jS4DEKIM3L9bK0wcyNoipBstjs6kS3G4ADyOFlSGAixaYWgwCXC-AANxIBlocEswHprUuxMV+GgAGEuiQoABzDXAOBKhAEMlA6mtKLW23hbQWmChFVqxm6-X4VLivWiro4ANB-0iwwR-34EVJRy6NGeVpIOMaMTmxxa5PY8wCEWpDPQM2OQsJhXC+OoRwAGSx+FT6YiJezJzzcALRZbWfLNcrhejEsLMfWuuUcAAHtg8PhfHBhCRDAQXpgFCAKCAIJgYA91MhQNoxF4AAraBD0ZAgEhKha+DcgABGODIAGs4DAmFowIszcgIuqm5dDA6AGAA6j08AJGQzAXskqBKskACeV5gPQl6bosjB4CeL5mugJDIIuBiMJusz0JOJQvmA76fvYtaLHARGqqRIDkZOTC-pUACKtAQPAzEkXAm4UthV6PiQj5IQY0APpgkSwGBqC+DAXTIAAHAADCJYiMGBL6YFe8lwNhNoPgAjnx8C4dul4oGMAC0UBwHA873pusiWWicC4SQ+GEUgxGsYw6CoP+OCASA9BcXAvH8UxgUscJICsI+SkqWpSAAEybhEJAFL+hriARV5QNATGbpwAAqkl2UFyVKuqACSUDzrATBgJEO5PK1LDSQl9UCAIQA)

**Input:**
```jsx
  const { firstName, lastName } = useMemo(
    () => parseFullName(fullName),
    [fullName],
  );
```

**Output:**
```jsx
  const { firstName, lastName } = useMemo(() => parseFullName(fullName), [
    fullName
  ]);
```

**Expected behavior:**
Input should be kept unchanged.

```jsx
  const { firstName, lastName } = useMemo(
    () => parseFullName(fullName),
    [fullName],
  );
```

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #7671

I don't usually write React very much. What do you think of this heuristic?

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by thorn0** [comment]
This works, but a cleaner solution would be to make the `shouldGroupLastArg` function return `false` in this case.

**Review by thorn0 — APPROVED** [review]

File: src/language-js/print/call-arguments.js [hunk 1]
```diff
@@ -289,4 +284,15 @@ function shouldGroupFirstArg(args) {
   );
 }
 
+function isReactHookCallWithDepsArray(args, isArrowBodyBlock) {
+  return (
+    args.length === 2 &&
+    args[0].type === "ArrowFunctionExpression" &&
+    getFunctionParameters(args[0]).length === 0 &&
```
Comment (fisker) [inline src/language-js/print/call-arguments.js:291]: Maybe we can allow some simple parameters, so we can format this case the same way

**Prettier pr-10238**
[Playground link](https://deploy-preview-10238--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEBtTAzASwCcNMBdTAXkwEMiiaBPAOiLgBMBXMOACgB0omTLwCUVAHw4aYMJwC2nADY0YEIgBpMcJXHkIYWmXMUq1m7bv2wyGwcOx5lSgHI19ZQaIDcgwZChSXEISLApqOgYWNi4eASERYwVlVXUtHT0DLXFKKWwk01SLDOsYW3scJyVXdzhPKB8QDRAIAAcYAnRkUEiIAHcABToENGQQGiU+plHmgCMGMABrOBgAZVaZAigAc2QYIk44Zr1ZjnYOABkaHc4abbgAMXV5VQ6dsZpONSaQAAsYeRKADqvwI8DQGx4qxGYIIADcwYwxmA0DMQFs0HAiDABgxti9kHgJpjmgArNAADwAQgtlmtahctnBCcSjiByRTVlttroAIqcCDwFlKEkgDYkLFjWY0U5KH6tIhbGBAgjsGC-ZAADgADM0FRBMUCGK0xgq4JiiHDmc0AI4C+C4tqjFA0NAAWigcA4HB+bDtxDguLuBKQRJFbMx8gIewOEe5fPtzNDrOaMBlKrVGqQACZUwwCEpuQBhCDyEMgc0AVh+nExABUZc6w6K4YcAJJQc6wVZgRXtACCndWMEYumFmIAvhOgA)
```sh
--parser babel
```

**Input:**
```jsx
const [ first ] = array.reduce(
  () => [accumulator, element, accumulator, element],
  [fullName]
);

const [ first ] = array.reduce(
  (accumulator, element, ) => [accumulator, element],
  [fullName]
);
```

**Output:**
```jsx
const [first] = array.reduce(
  () => [accumulator, element, accumulator, element],
  [fullName]
);

const [first] = array.reduce((accumulator, element) => [accumulator, element], [
  fullName,
]);

```

Comment (thorn0) [reply src/language-js/print/call-arguments.js:291]: Let's just prevent hugging of the last argument if:
```js
args.length === 2 &&
args[0].type === "ArrowFunctionExpression" &&
args[1].type === "ArrayExpression"
```
It's difficult for me to imagine a case when last arg hugging would look good in such a call.

**Review by thorn0 — APPROVED** [review]

**Review by lipis — APPROVED** [review]

**Comment by DScheglov** [comment]
Hey, @thorn0 , sorry for probably silly question.
Which `prettier` version receives this update?

**Comment by vexii** [comment]
could we get a option to disable this?

**Comment by pyros2097** [comment]
Yep can we get an option to disable this. My code looks very hard to read now,
```js
const {
  data: { items },
} = useTodos({ limit: 10 });
```

**Pull request merged**